pppd: Fill in default gateway on Linux

On Linux, when pppd creates the default route, it does not set the peer
address as gateway, leading to a default route without gateway address.

This behaviour breaks various downstream programs which attempt to infer
the default gateway IP address from the system default route entry.

This patch addresses the issue by filling in the peer address as gateway
when generating the default route entry.

Signed-off-by: Jo-Philipp Wich <jo@mein.io>

--- a/pppd/sys-linux.c
+++ b/pppd/sys-linux.c
<<<<<<< HEAD
@@ -2247,6 +2247,9 @@ int sifdefaultroute (int unit, u_int32_t
=======
@@ -1717,6 +1717,9 @@ int sifdefaultroute (int unit, u_int32_t
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
     memset (&rt, 0, sizeof (rt));
     SET_SA_FAMILY (rt.rt_dst, AF_INET);
 
+    SET_SA_FAMILY(rt.rt_gateway, AF_INET);
+    SIN_ADDR(rt.rt_gateway) = gateway;
+
     rt.rt_dev = ifname;
     rt.rt_metric = dfl_route_metric + 1; /* +1 for binary compatibility */
 
<<<<<<< HEAD
@@ -2255,7 +2258,7 @@ int sifdefaultroute (int unit, u_int32_t
=======
@@ -1725,7 +1728,7 @@ int sifdefaultroute (int unit, u_int32_t
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 	SIN_ADDR(rt.rt_genmask) = 0L;
     }
 
-    rt.rt_flags = RTF_UP;
+    rt.rt_flags = RTF_UP | RTF_GATEWAY;
     if (ioctl(sock_fd, SIOCADDRT, &rt) < 0) {
<<<<<<< HEAD
 	if ( ! ok_error ( errno ))
=======
 	if (!ok_error(errno))
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 	    error("default route ioctl(SIOCADDRT): %m");
